android/app/src/main/java/com/expensify/chat/CustomNotificationProvider.java

```diff
@@ -0,0 +1,234 @@
+        }
+
+        IconCompat iconCompat = fetchIcon(avatar, FALLBACK_ICON_ID);
+        Person person = new Person.Builder()
+                .setIcon(iconCompat)
+                .setKey(accountID)
+                .setName(name)
+                .build();
+
+        if (!notificationCache.people.containsKey(accountID)) {
+            notificationCache.people.put(accountID, person);
+        }
+
+        notificationCache.messages.add(new NotificationCache.Message(person, message, time));
+
+        NotificationCompat.MessagingStyle messagingStyle = new NotificationCompat.MessagingStyle(person)
+                .setGroupConversation(notificationCache.people.size() > 2 || !roomName.isEmpty())
+                .setConversationTitle(conversationTitle);
+
+        for (NotificationCache.Message cachedMessage : notificationCache.messages) {
+            messagingStyle.addMessage(cachedMessage.text, cachedMessage.time, cachedMessage.person);
+        }
+
+        if (notificationCache.prevNotificationID != -1) {
+            NotificationManagerCompat.from(context).cancel(notificationCache.prevNotificationID);
+        }
+
+        builder.setStyle(messagingStyle);
+        notificationCache.prevNotificationID = notificationID;
+    }
+
+    /**
+     * Check if we are showing a notification related to a reportID.
+     * If not, create a new NotificationCache so we can build a conversation notification
+     * as the messages come.
+     *
+     * @param reportID Report ID.
+     * @return Notification Cache.
+     */
+    private NotificationCache findOrCreateNotificationCache(int reportID) {
+        NotificationCache notificationCache = cache.get(reportID);
+
+        if (notificationCache == null) {
+            notificationCache = new NotificationCache();
+            cache.put(reportID, notificationCache);
+        }
+
+        return notificationCache;
+    }
+
+    /**
+     * Remove the notification data from the cache when the user dismisses the notification.
+     *
+     * @param message Push notification's message
+     */
+    public void onDismissNotification(PushMessage message) {
+        try {
+            JsonMap payload = JsonValue.parseString(message.getExtra(PAYLOAD_KEY)).optMap();
+            int reportID = payload.get("reportID").getInt(-1);
+
+            if (reportID == -1) {
+                return;
+            }
+
+            cache.remove(reportID);
+        } catch (Exception e) {
+            Log.e(TAG, "Failed to delete conversation cache");
+        }
+    }
+
+    @NonNull
+    private IconCompat fetchIcon(@NonNull String urlString, @DrawableRes int fallbackId) {
+        // TODO: Add disk LRU cache
+
+        URL parsedUrl = null;
+        try {
+            parsedUrl = urlString == null ? null : new URL(urlString);
+        } catch (MalformedURLException e) {
+            Log.e(TAG, "Failed to resolve URL " + urlString, e);
+        }
+
+        if (parsedUrl != null) {
```

Comment:
Maybe return early instead, if parsedURl is null?

Reply:
Done!